include/tscpp/api/HttpMsgComp.h

```diff
@@ -0,0 +1,449 @@
+  MimeField &operator=(const MimeField &) = delete;
+
+  MimeField(MimeField &&source);
+
+  // Move assign only allowed if the two instances refer to the same message.
+  //
+  MimeField &operator=(MimeField &&source);
+
+  void destroy();
+
+  // Next field, returns invalid instance if none.
+  //
+  MimeField next() const;
+
+  // Next field with same name, return invalid instance if none.
+  //
+  MimeField nextDup() const;
+
+  std::string_view nameGet() const;
+
+  void nameSet(std::string_view);
+
+  void valuesClear();
+
+  // Returns number of values. Values index are from 0 to valuesCount() - 1.
+  //
+  int valuesCount() const;
+
+  std::string_view valGet(int idx) const;
+
+  void valSet(int idx, std::string_view new_value);
+
+  // Insert a new value at index idx.  All values with index >= idx prior to calling this have their index incremented by one.
+  //
+  void valInsert(int idx, std::string_view new_value);
+
+  // Append a new value at the end.
+  //
+  void valAppend(std::string_view new_value);
+
+  ~MimeField();
+
+private:
+  MsgBase &_msg;
+  TSMLoc _loc;
+};
+
+class MsgBuffer : noncopyable
+{
+public:
+  MsgBuffer(TSMBuffer hndl) : _hndl(hndl) {}
+
+  // Returns the handle.
+  //
+  TSMBuffer
+  bufHndl() const
+  {
+    return _hndl;
+  }
+
+protected:
+  MsgBuffer() {}
+
+  void
+  set(TSMBuffer hndl)
+  {
+    _hndl = hndl;
+  }
+
+private:
+  TSMBuffer _hndl;
+};
+
+class MsgBase : public MsgBuffer
+{
+public:
+  MsgBase(TSMBuffer bHndl, TSMLoc msgLoc) : MsgBuffer(bHndl), _msgLoc(msgLoc) {}
+
+  TSMLoc
+  msgLoc() const
+  {
+    return _msgLoc;
+  }
+
+  bool
+  hasMsg() const
+  {
+    return msgLoc() != TS_NULL_MLOC;
+  }
+
+  //  Returns number of mime header lines in HTTP message.  Can only be called if hasMsg() is true.
+  //
+  int
+  mimeFieldsCount() const
+  {
+    assert(hasMsg());
+
+    return TSMimeHdrFieldsCount(bufHndl(), msgLoc());
+  }
+
+  MimeField
+  newMimeField()
+  {
+    assert(hasMsg());
+
+    TSMLoc loc;
+    if (TSMimeHdrFieldCreate(bufHndl(), msgLoc(), &loc) != TS_SUCCESS) {
+      loc = TS_NULL_MLOC;
+    }
+    return MimeField(*this, loc);
+  }
+
+  ~MsgBase()
+  {
+    if (_msgLoc && bufHndl()) {
+      assert(TSHandleMLocRelease(bufHndl(), TS_NULL_MLOC, _msgLoc) == TS_SUCCESS);
+    }
+  }
+
+protected:
+  MsgBase() : MsgBuffer(nullptr), _msgLoc(TS_NULL_MLOC) {}
+
+  void
+  setMsgLoc(TSMLoc msgLoc)
+  {
+    _msgLoc = msgLoc;
+  }
+
+private:
+  TSMLoc _msgLoc;
+};
+
+template <TSReturnCode (*getInTxn)(TSHttpTxn txnp, TSMBuffer *, TSMLoc *)> class TxnMsg : public MsgBase
+{
+public:
+  TxnMsg() : MsgBase(nullptr, TS_NULL_MLOC) {}
+
+  // Acutal type of txnHndl must be TSHttpTxn.  Using void * for compatability with
+  // atscppapi::Transaction::getAtsHandle().
+  //
+  explicit TxnMsg(void *txnHndl) { _init(txnHndl); }
+
+  // Must not be called on instance that already has a message.
+  //
+  bool
+  init(void *txnHndl)
+  {
+    assert(!hasMsg());
```

Comment:
Seems obey the ATS Coding Style. See here https://cwiki.apache.org/confluence/display/TS/Coding+Style

Reply:
Can you clarify?  I don't find any mention of assert in the Coding Style.